File: src/JavaScript/sync_voting_summary.js
// sync_voting_summary.js
// Sync voting summary from Koios → PostgreSQL
// Equivalent to sync_voting_summary.py
//
// Koios: GET /api/v1/proposal_voting_summary?_proposal_id={id}
// Target: proposal_voting_summary table
// Reads proposal_ids from PostgreSQL proposals table.

require('dotenv').config();
const fetch = require('node-fetch');
const crypto = require('crypto');
const { pgConnect, pgUpsertBatch, info, warn, error } = require('./helpers');
const { TABLE_COLUMNS } = require('./config');

const DATABASE_URL = process.env.DATABASE_URL;
if (!DATABASE_URL) {
  console.error('❌ Thiếu DATABASE_URL trong .env');
  process.exit(1);
}

const KOIOS_BASE = 'https://api.koios.rest/api/v1';
const BATCH_SIZE = 25;
const API_DELAY = 500;

const COLUMNS = TABLE_COLUMNS.proposal_voting_summary;

async function jsonGet(url, headers = {}) {
  const res = await fetch(url, { headers });
  if (!res.ok) throw new Error(`${url} → HTTP ${res.status}`);
  return res.json();
}

async function koiosGet(endpoint, params = {}) {
  const url = new URL(`${KOIOS_BASE}/${endpoint}`);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  return jsonGet(url.toString(), { Accept: 'application/json' });
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function genUuid() {
  return crypto.randomUUID();
}

function nowIso() {
  return new Date().toISOString();
}

async function syncVotingSummary() {
  info('=== Sync: proposal_voting_summary (Koios → PostgreSQL) ===');

  const client = await pgConnect();

  try {
    // 1. Get active proposal IDs
    const propRes = await client.query(
      "SELECT proposal_id FROM proposals WHERE status IN ('voting','active') ORDER BY proposal_id"
    );
    let proposalIds = propRes.rows.map(r => r.proposal_id);
    proposalIds = proposalIds.filter(pid => !pid.startsWith('final_verify_v3'));
    info(`[voting_summary] Found ${proposalIds.length} active proposals`);

    let allRows = [];
    let errors = 0;

    for (let i = 0; i < proposalIds.length; i++) {
      const pid = proposalIds[i];
      try {
        const data = await koiosGet('proposal_voting_summary', { _proposal_id: pid });
        if (!data || !Array.isArray(data) || data.length === 0) continue;

        const d = data[0];
        allRows.push({
          id: genUuid(),
          proposal_id: pid,
          proposal_type: d.proposal_type,
          epoch_no: d.epoch_no,
          drep_yes_votes_cast: d.drep_yes_votes_cast || 0,
          drep_active_yes_vote_power: d.drep_active_yes_vote_power || 0,
          drep_yes_vote_power: d.drep_yes_vote_power || 0,
          drep_yes_pct: d.drep_yes_pct || 0,
          drep_no_votes_cast: d.drep_no_votes_cast || 0,
          drep_active_no_vote_power: d.drep_active_no_vote_power || 0,
          drep_no_vote_power: d.drep_no_vote_power || 0,
          drep_no_pct: d.drep_no_pct || 0,
          drep_abstain_votes_cast: d.drep_abstain_votes_cast || 0,
          drep_active_abstain_vote_power: d.drep_active_abstain_vote_power || 0,
          drep_abstain_vote_power: d.drep_abstain_vote_power || 0,
          drep_always_abstain_vote_power: d.drep_always_abstain_vote_power || 0,
          drep_always_no_confidence_vote_power: d.drep_always_no_confidence_vote_power || 0,
          pool_yes_votes_cast: d.pool_yes_votes_cast || 0,
          pool_active_yes_vote_power: d.pool_active_yes_vote_power || 0,
          pool_yes_vote_power: d.pool_yes_vote_power || 0,
          pool_yes_pct: d.pool_yes_pct || 0,
          pool_no_votes_cast: d.pool_no_votes_cast || 0,
          pool_active_no_vote_power: d.pool_active_no_vote_power || 0,
          pool_no_vote_power: d.pool_no_vote_power || 0,
          pool_no_pct: d.pool_no_pct || 0,
          pool_abstain_votes_cast: d.pool_abstain_votes_cast || 0,
          pool_active_abstain_vote_power: d.pool_active_abstain_vote_power || 0,
          pool_passive_always_abstain_votes_assigned: d.pool_passive_always_abstain_votes_assigned || 0,
          pool_passive_always_abstain_vote_power: d.pool_passive_always_abstain_vote_power || 0,
          pool_passive_always_no_confidence_votes_assigned: d.pool_passive_always_no_confidence_votes_assigned || 0,
          pool_passive_always_no_confidence_vote_power: d.pool_passive_always_no_confidence_vote_power || 0,
          committee_yes_votes_cast: d.committee_yes_votes_cast || 0,
          committee_yes_pct: d.committee_yes_pct || 0,
          committee_no_votes_cast: d.committee_no_votes_cast || 0,
          committee_no_pct: d.committee_no_pct || 0,
          committee_abstain_votes_cast: d.committee_abstain_votes_cast || 0,
          data_fetched_at: nowIso(),
          created_at: nowIso(),
          updated_at: nowIso(),
        });

        // Flush batch
        if (allRows.length >= BATCH_SIZE) {
          try {
            await pgUpsertBatch(client, 'proposal_voting_summary', COLUMNS, allRows, {
              conflict_cols: ['proposal_id'],
              preserve_cols: ['id', 'created_at'],
              do_update: true,
            });
          } catch (e) {
            error(`[voting_summary] Insert error: ${e.message}`);
          }
          allRows = [];
        }
      } catch (e) {
        errors++;
        if (errors <= 5) warn(`[voting_summary] Error for ${pid}: ${e.message}`);
      }

      if ((i + 1) % 10 === 0 || i === proposalIds.length - 1) {
        process.stdout.write(`\r  [voting_summary] ${i + 1}/${proposalIds.length}, errors=${errors}`);
      }
      await sleep(API_DELAY);
    }

    // Flush remaining
    if (allRows.length > 0) {
      try {
        await pgUpsertBatch(client, 'proposal_voting_summary', COLUMNS, allRows, {
          conflict_cols: ['proposal_id'],
          preserve_cols: ['id', 'created_at'],
          do_update: true,
        });
      } catch (e) {
        error(`[voting_summary] Final insert error: ${e.message}`);
      }
    }

    info(`\n[voting_summary] Done, errors=${errors}`);
    const cntRes = await client.query('SELECT count(*)::int as cnt FROM proposal_voting_summary');
    info(`[voting_summary] Row count: ${cntRes.rows[0].cnt}`);
    return cntRes.rows[0].cnt;
  } finally {
    await client.end();
  }
}

if (require.main === module) {
  syncVotingSummary().catch(err => {
    console.error(err);
    process.exit(1);
  });
}

module.exports = { syncVotingSummary };

File: scripts/python/sync_voting_summary.py
"""Sync voting summary from Koios → Neon + Supabase (dual-write).

Koios: GET /api/v1/proposal_voting_summary?_proposal_id={id}
Target: proposal_voting_summary table
Reads proposal_ids from Neon proposals table.
"""

import sys
import time

from config import TABLE_COLUMNS, BATCH_SIZE, API_DELAY
from helpers import (
    get_logger, check_env, koios_get, neon_connect,
    neon_truncate, neon_upsert_batch, neon_row_count, now_iso, gen_uuid,
    supabase_upsert_batch, supabase_row_count,
)


def sync_voting_summary(logger=None):
    if logger is None:
        logger = get_logger()

    logger.info("=== Sync: proposal_voting_summary (Koios → Neon) ===")
    columns = TABLE_COLUMNS["proposal_voting_summary"]

    # 1. Get active proposal IDs from Neon
    conn = neon_connect()
    with conn.cursor() as cur:
        cur.execute(
            "SELECT proposal_id FROM proposals "
            "WHERE status IN ('voting','active') ORDER BY proposal_id"
        )
        proposal_ids = [r[0] for r in cur.fetchall()]
    # Skip test proposals
    proposal_ids = [pid for pid in proposal_ids if not pid.startswith("final_verify_v3")]
    logger.info(f"[voting_summary] Found {len(proposal_ids)} active proposals")

    # 2. Upsert (no truncate)
    conn = neon_connect()

    # 3. Fetch voting summary for each proposal
    all_rows = []
    errors = 0

    for i, pid in enumerate(proposal_ids):
        try:
            data = koios_get("proposal_voting_summary", params={"_proposal_id": pid})
            if not data or not isinstance(data, list) or len(data) == 0:
                continue

            d = data[0]
            all_rows.append({
                "id": gen_uuid(),
                "proposal_id": pid,
                "proposal_type": d.get("proposal_type"),
                "epoch_no": d.get("epoch_no"),
                "drep_yes_votes_cast": d.get("drep_yes_votes_cast", 0),
                "drep_active_yes_vote_power": d.get("drep_active_yes_vote_power", 0),
                "drep_yes_vote_power": d.get("drep_yes_vote_power", 0),
                "drep_yes_pct": d.get("drep_yes_pct", 0),
                "drep_no_votes_cast": d.get("drep_no_votes_cast", 0),
                "drep_active_no_vote_power": d.get("drep_active_no_vote_power", 0),
                "drep_no_vote_power": d.get("drep_no_vote_power", 0),
                "drep_no_pct": d.get("drep_no_pct", 0),
                "drep_abstain_votes_cast": d.get("drep_abstain_votes_cast", 0),
                "drep_active_abstain_vote_power": d.get("drep_active_abstain_vote_power", 0),
                "drep_abstain_vote_power": d.get("drep_abstain_vote_power", 0),
                "drep_always_abstain_vote_power": d.get("drep_always_abstain_vote_power", 0),
                "drep_always_no_confidence_vote_power": d.get("drep_always_no_confidence_vote_power", 0),
                "pool_yes_votes_cast": d.get("pool_yes_votes_cast", 0),
                "pool_active_yes_vote_power": d.get("pool_active_yes_vote_power", 0),
                "pool_yes_vote_power": d.get("pool_yes_vote_power", 0),
                "pool_yes_pct": d.get("pool_yes_pct", 0),
                "pool_no_votes_cast": d.get("pool_no_votes_cast", 0),
                "pool_active_no_vote_power": d.get("pool_active_no_vote_power", 0),
                "pool_no_vote_power": d.get("pool_no_vote_power", 0),
                "pool_no_pct": d.get("pool_no_pct", 0),
                "pool_abstain_votes_cast": d.get("pool_abstain_votes_cast", 0),
                "pool_active_abstain_vote_power": d.get("pool_active_abstain_vote_power", 0),
                "pool_passive_always_abstain_votes_assigned": d.get("pool_passive_always_abstain_votes_assigned", 0),
                "pool_passive_always_abstain_vote_power": d.get("pool_passive_always_abstain_vote_power", 0),
                "pool_passive_always_no_confidence_votes_assigned": d.get("pool_passive_always_no_confidence_votes_assigned", 0),
                "pool_passive_always_no_confidence_vote_power": d.get("pool_passive_always_no_confidence_vote_power", 0),
                "committee_yes_votes_cast": d.get("committee_yes_votes_cast", 0),
                "committee_yes_pct": d.get("committee_yes_pct", 0),
                "committee_no_votes_cast": d.get("committee_no_votes_cast", 0),
                "committee_no_pct": d.get("committee_no_pct", 0),
                "committee_abstain_votes_cast": d.get("committee_abstain_votes_cast", 0),
                "data_fetched_at": now_iso(),
                "created_at": now_iso(),
                "updated_at": now_iso(),
            })

            # Insert in batches
            if len(all_rows) >= BATCH_SIZE:
                try:
                    neon_upsert_batch(conn, "proposal_voting_summary", columns, all_rows, conflict_cols=["proposal_id"], preserve_cols=("id", "created_at"))
                except Exception as e:
                    logger.error(f"[voting_summary] Neon insert error: {e}")
                try:
                    supabase_upsert_batch("proposal_voting_summary", columns, all_rows, conflict_cols=["proposal_id"], omit_cols=("id",))
                except Exception as e:
                    logger.error(f"[voting_summary] Supabase insert error: {e}")
                all_rows = []

        except Exception as e:
            errors += 1
            if errors <= 5:
                logger.warning(f"[voting_summary] Error for {pid}: {e}")

        if (i + 1) % 10 == 0 or i == len(proposal_ids) - 1:
            sys.stdout.write(f"\r  [voting_summary] {i+1}/{len(proposal_ids)}, errors={errors}")
            sys.stdout.flush()
        time.sleep(API_DELAY)

    # Insert remaining
    if all_rows:
        try:
            neon_upsert_batch(conn, "proposal_voting_summary", columns, all_rows, conflict_cols=["proposal_id"], preserve_cols=("id", "created_at"))
        except Exception as e:
            logger.error(f"[voting_summary] Neon final insert error: {e}")
        try:
            supabase_upsert_batch("proposal_voting_summary", columns, all_rows, conflict_cols=["proposal_id"], omit_cols=("id",))
        except Exception as e:
            logger.error(f"[voting_summary] Supabase final insert error: {e}")

    logger.info(f"\n[voting_summary] Done, errors={errors}")
    count = neon_row_count(conn, "proposal_voting_summary")
    logger.info(f"[voting_summary] Neon row count: {count}")
    conn.close()
    try:
        sb_count = supabase_row_count("proposal_voting_summary")
        logger.info(f"[voting_summary] Supabase row count: {sb_count}")
    except Exception as e:
        logger.warning(f"[voting_summary] Supabase count failed: {e}")
    return count


if __name__ == "__main__":
    check_env()
    sync_voting_summary()
